k-distribution/k-tutorial/1_basic/19_debugging/README.md

```diff
@@ -1 +1,359 @@
+
+Breakpoint 1, LESSON-19-C.isEven.rhs () at /home/dwightguth/kframework-5.0.0/k-distribution/k-tutorial/1_basic/19_debugging/lesson-19-c.k:6
+6         rule [isEven]: isEven(I) => true requires I %Int 2 ==Int 0
+(gdb)
+```
+
+We can also set a breakpoint for when a rule's side condition is evaluated
+by means of the `MODULE-NAME.label.sc` syntax:
+
+```
+(gdb) break LESSON-19-C.isEven.sc
+Breakpoint 1 at 0x2afd70: file /home/dwightguth/kframework-5.0.0/k-distribution/k-tutorial/1_basic/19_debugging/lesson-19-c.k, line 6.
+(gdb) run
+Starting program: /home/dwightguth/kframework-5.0.0/k-distribution/k-tutorial/1_basic/19_debugging/lesson-19-c-kompiled/interpreter .krun-2021-08-13-14-41-48-1BoGfJRbYc/tmp.in.kg4F8cwfCe -1 .krun-2021-08-13-14-41-48-1BoGfJRbYc/result.kore
+[Thread debugging using libthread_db enabled]
+Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
+
+Breakpoint 1, LESSON-19-C.isEven.sc (VarI=4) at /home/dwightguth/kframework-5.0.0/k-distribution/k-tutorial/1_basic/19_debugging/lesson-19-c.k:6
+6         rule [isEven]: isEven(I) => true requires I %Int 2 ==Int 0
+(gdb) finish
+Run till exit from #0  LESSON-19-C.isEven.sc (VarI=4) at /home/dwightguth/kframework-5.0.0/k-distribution/k-tutorial/1_basic/19_debugging/lesson-19-c.k:6
+0x00000000002b2662 in LblisEven'LParUndsRParUnds'LESSON-19-C'Unds'Bool'Unds'Int (_1=4) at /home/dwightguth/kframework-5.0.0/k-distribution/k-tutorial/1_basic/19_debugging/lesson-19-c.k:5
+5         syntax Bool ::= isEven(Int) [function]
+Value returned is $1 = true
+(gdb)
+```
+
+Here we have used the built-in GDB command `finish` to tell us whether the
+side condition returned true or not. Note that once again, we see the
+substitution that was matched from the left-hand side. Like before, a variable
+will only appear here if it is used in the side condition.
+
+## Debugging rule matching
+
+Sometimes it is useful to try to determine why a particular rule did or did
+not apply. K provides some basic debugging commands which make it easier
+to determine this.
+
+Consider the following K definition (`lesson-19-d.k`):
+
+```k
+module LESSON-19-D
+
+  syntax Foo ::= foo(Bar)
+  syntax Bar ::= bar(Baz) | bar2(Baz)
+  syntax Baz ::= baz() | baz2()
+
+  rule [baz]: foo(bar(baz())) => .K
+
+endmodule
+```
+
+Suppose we try to run the program `foo(bar(baz2()))`. It is obvious from this
+example why the rule in this definition will not apply. However, in practice,
+such cases are not always obvious. You might look at a rule and not immediately
+spot why it didn't apply on a particular term. For this reason, it can be
+useful to get the debugger to provide a log about how it tried to match that
+term. You can do this with the `k match` command. If you are stopped after
+having run `k start` or `k step`, you can obtain this log by running the
+command `k match MODULE.label subject` for a particular top-level rule label.
+
```

Comment:
I'm not 100% sure I understand this functionality. So if I'm anywhere in the execution, I can write `k match LESSON-19-D.baz subject` and see whether or not my current configuration matches a given rule? Or does it only work when the program is stuck?

Also, for clarity, I'd add the prompt here at the top of the output to show an example of what exactly the input is:
```
(gdb) k match LESSON-19-D.baz subject
Subject:
baz2 ( )
does not match pattern:
baz ( )
```